Spurious test failure: "error: linking with `cc` failed"

```
error: compilation failed!

status: exit code: 101

command: target/release/miri tests/run-pass/aux_test.rs -L /tmp --target=x86_64-unknown-linux-gnu -Z dump-mir=all -Z mir-opt-level=3 -Z dump-mir-dir=/tmp/aux_test -L /tmp/aux_test.stage-id.mir-opt.libaux -C prefer-dynamic -o /tmp/aux_test.stage-id -Zmir-opt-level=0 -Zmir-emit-validate=1

stdout:

------------------------------------------

------------------------------------------

stderr:

------------------------------------------

error: linking with `cc` failed: exit code: 1

  |

  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/aux_test.0.o" "-o" "/tmp/aux_test.stage-id" "/tmp/aux_test.crate.allocator.o" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-L" "/tmp" "-L" "/tmp/aux_test.stage-id.mir-opt.libaux" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/tmp/aux_test.stage-id.mir-opt.libaux" "-l" "dep" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-l" "std-ab8512bb715ab82f" "-Wl,-Bstatic" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-53095306c4bf2e4a.rlib" "-Wl,-Bdynamic" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util"

  = note: /usr/bin/ld: cannot find /tmp/aux_test.0.o: No such file or directory

          /usr/bin/ld: cannot find /tmp/aux_test.crate.allocator.o: No such file or directory

          collect2: ld returned 1 exit status
```
See <https://travis-ci.org/solson/miri/builds/265176951> for an example. Re-running the test usually fixes this. I seem to remember I even saw that locally.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -37,6 +37,7 @@ cargo_miri = ["cargo_metadata"]
 
 [dev-dependencies]
 compiletest_rs = "0.2.6"
+tempdir = "0.3"
 
 [workspace]
 exclude = ["xargo", "cargo-miri-test", "rustc_tests"]

--- a/tests/compiletest.rs
+++ b/tests/compiletest.rs
@@ -37,9 +37,9 @@ fn compile_fail(sysroot: &Path, path: &str, target: &str, host: &str, fullmir: b
         path,
         target
     );
+    let build_dir = TempDir::new("miri-tests").unwrap();
     let mut config = compiletest::Config::default();
     config.mode = "compile-fail".parse().expect("Invalid mode");
-    let build_dir = TempDir::new("miri-tests").unwrap();
     config.build_base = build_dir.path().to_owned();
     config.rustc_path = miri_path();
     let mut flags = Vec::new();
@@ -70,9 +70,9 @@ fn compile_fail(sysroot: &Path, path: &str, target: &str, host: &str, fullmir: b
 
 fn run_pass(path: &str) {
     eprintln!("## Running run-pass tests in {} against rustc", path);
+    let build_dir = TempDir::new("miri-tests").unwrap();
     let mut config = compiletest::Config::default();
     config.mode = "run-pass".parse().expect("Invalid mode");
-    let build_dir = TempDir::new("miri-tests").unwrap();
     config.build_base = build_dir.path().to_owned();
     config.src_base = PathBuf::from(path);
     if let Some(rustc_path) = rustc_test_suite() {
@@ -95,9 +95,9 @@ fn miri_pass(path: &str, target: &str, host: &str, fullmir: bool, opt: bool) {
         target,
         opt_str
     );
+    let build_dir = TempDir::new("miri-tests").unwrap();
     let mut config = compiletest::Config::default();
     config.mode = "mir-opt".parse().expect("Invalid mode");
-    let build_dir = TempDir::new("miri-tests").unwrap();
     config.build_base = build_dir.path().to_owned();
     config.src_base = PathBuf::from(path);
     config.target = target.to_owned();

--- a/tests/compiletest.rs
+++ b/tests/compiletest.rs
@@ -1,10 +1,12 @@
 #![feature(slice_concat_ext)]
 
 extern crate compiletest_rs as compiletest;
+extern crate tempdir;
 
 use std::slice::SliceConcatExt;
 use std::path::{PathBuf, Path};
 use std::io::Write;
+use tempdir::TempDir;
 
 macro_rules! eprintln {
     ($($arg:tt)*) => {
@@ -35,8 +37,10 @@ fn compile_fail(sysroot: &Path, path: &str, target: &str, host: &str, fullmir: b
         path,
         target
     );
-    let mut config = compiletest::default_config();
+    let mut config = compiletest::Config::default();
     config.mode = "compile-fail".parse().expect("Invalid mode");
+    let build_dir = TempDir::new("miri-tests").unwrap();
+    config.build_base = build_dir.path().to_owned();
     config.rustc_path = miri_path();
     let mut flags = Vec::new();
     if rustc_test_suite().is_some() {
@@ -66,8 +70,10 @@ fn compile_fail(sysroot: &Path, path: &str, target: &str, host: &str, fullmir: b
 
 fn run_pass(path: &str) {
     eprintln!("## Running run-pass tests in {} against rustc", path);
-    let mut config = compiletest::default_config();
+    let mut config = compiletest::Config::default();
     config.mode = "run-pass".parse().expect("Invalid mode");
+    let build_dir = TempDir::new("miri-tests").unwrap();
+    config.build_base = build_dir.path().to_owned();
     config.src_base = PathBuf::from(path);
     if let Some(rustc_path) = rustc_test_suite() {
         config.rustc_path = rustc_path;
@@ -89,8 +95,10 @@ fn miri_pass(path: &str, target: &str, host: &str, fullmir: bool, opt: bool) {
         target,
         opt_str
     );
-    let mut config = compiletest::default_config();
+    let mut config = compiletest::Config::default();
     config.mode = "mir-opt".parse().expect("Invalid mode");
+    let build_dir = TempDir::new("miri-tests").unwrap();
+    config.build_base = build_dir.path().to_owned();
     config.src_base = PathBuf::from(path);
     config.target = target.to_owned();
     config.host = host.to_owned();

